// @bayajidalam/apollo-cli/dist/commands/generate.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.generateCommand = void 0;
const chalk_1 = __importDefault(require("chalk"));
const commander_1 = require("commander");
const inquirer_1 = __importDefault(require("inquirer"));
const path_1 = __importDefault(require("path"));
const file_generator_1 = require("../utils/file-generator");
exports.generateCommand = new commander_1.Command('generate')
    .alias('g')
    .description('Generate a new resource')
    .argument('[type]', 'Type of resource to generate (e.g., module)')
    .argument('[name]', 'Name of the resource')
    .addHelpText('after', `
${chalk_1.default.yellow('Examples:')}
  ${chalk_1.default.cyan('$ apollo-cli generate module user')}
  ${chalk_1.default.cyan('$ apollo-cli g module product')}
`)
    .action(async (type, name) => {
    if (!type) {
        const answers = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'type',
                message: chalk_1.default.blue('What would you like to generate?'),
                choices: ['module'],
            },
        ]);
        type = answers.type;
    }
    if (type !== 'module') {
        console.error(chalk_1.default.red('❌ Only "module" generation is currently supported.'));
        return;
    }
    if (!name) {
        const answers = await inquirer_1.default.prompt([
            {
                type: 'input',
                name: 'name',
                message: chalk_1.default.blue('What is the name of the module?'),
                validate: (input) => input.trim() === '' ? 'Module name is required' : true,
            },
        ]);
        name = answers.name;
    }
    const moduleNameRaw = name.trim();
    const capitalize = (s) => s.charAt(0).toUpperCase() + s.slice(1);
    const camelize = (s) => s.charAt(0).toLowerCase() + s.slice(1);
    const moduleNamePascal = capitalize(moduleNameRaw);
    const moduleNameCamel = camelize(moduleNameRaw);
    console.log(chalk_1.default.cyan(`\n🚀 Generating module: ${chalk_1.default.bold(moduleNamePascal)}...\n`));
    // Assumes the command is run from the project root
    const baseDir = path_1.default.join(process.cwd(), 'src/modules', moduleNamePascal);
    const files = [
        {
            template: 'module/controller.ejs',
            output: `${moduleNameCamel}.controller.ts`,
        },
        {
            template: 'module/service.ejs',
            output: `${moduleNameCamel}.service.ts`,
        },
        { template: 'module/route.ejs', output: `${moduleNameCamel}.route.ts` },
        {
            template: 'module/interface.ejs',
            output: `${moduleNameCamel}.interface.ts`,
        },
        {
            template: 'module/validation.ejs',
            output: `${moduleNameCamel}.validation.ts`,
        },
        {
            template: 'module/constant.ejs',
            output: `${moduleNameCamel}.constant.ts`,
        },
    ];
    for (const file of files) {
        await (0, file_generator_1.renderTemplate)(file.template, path_1.default.join(baseDir, file.output), {
            moduleName: moduleNamePascal,
            camelModuleName: moduleNameCamel,
        });
    }
    console.log(chalk_1.default.green('\n✨ Module generated successfully! 🚀\n'));
});
