// @bayajidalam/apollo-cli/dist/commands/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.initCommand = void 0;
const chalk_1 = __importDefault(require("chalk"));
const child_process_1 = require("child_process");
const commander_1 = require("commander");
const fs_extra_1 = __importDefault(require("fs-extra"));
const inquirer_1 = __importDefault(require("inquirer"));
const path_1 = __importDefault(require("path"));
const file_generator_1 = require("../utils/file-generator");
// Helper function to get the latest version of a package from npm
async function getLatestVersion(packageName) {
    try {
        const version = (0, child_process_1.execSync)(`npm view ${packageName} version`, {
            encoding: 'utf-8',
            stdio: ['pipe', 'pipe', 'pipe'],
        }).trim();
        return `^${version}`;
    }
    catch (error) {
        console.warn(chalk_1.default.yellow(`⚠️  Could not fetch latest version for ${packageName}, using fallback`));
        return 'latest';
    }
}
// Fetch latest versions for all dependencies
async function getLatestDependencies() {
    console.log(chalk_1.default.blue('📦 Fetching latest package versions...'));
    const deps = [
        'express',
        'cors',
        'dotenv',
        'http-status',
        '@prisma/client',
        '@prisma/adapter-pg',
        'pg',
        'zod',
        'bcrypt',
        'jsonwebtoken',
        'cookie-parser',
    ];
    const devDeps = [
        'typescript',
        'ts-node-dev',
        '@types/node',
        '@types/express',
        '@types/cookie-parser',
        '@types/bcrypt',
        '@types/jsonwebtoken',
        '@types/pg',
        '@types/cors',
        'prisma',
        'eslint',
        'prettier',
    ];
    const [dependencies, devDependencies] = await Promise.all([
        Promise.all(deps.map(async (dep) => [dep, await getLatestVersion(dep)])),
        Promise.all(devDeps.map(async (dep) => [dep, await getLatestVersion(dep)])),
    ]);
    return {
        dependencies: Object.fromEntries(dependencies),
        devDependencies: Object.fromEntries(devDependencies),
    };
}
// Detect available package managers
function detectPackageManagers() {
    const all = ['npm', 'pnpm', 'yarn'];
    const available = [];
    all.forEach((pm) => {
        try {
            (0, child_process_1.execSync)(`${pm} --version`, { stdio: 'ignore' });
            available.push(pm);
        }
        catch {
            // Package manager not available
        }
    });
    return { available, all };
}
// Check if a package manager is installed
function isPackageManagerInstalled(pm) {
    try {
        (0, child_process_1.execSync)(`${pm} --version`, { stdio: 'ignore' });
        return true;
    }
    catch {
        return false;
    }
}
exports.initCommand = new commander_1.Command('init')
    .description('Initialize a new Apollo Gears project')
    .argument('[projectName]', 'Name of the project directory')
    .addHelpText('after', `
${chalk_1.default.yellow('Examples:')}
  ${chalk_1.default.cyan('$ apollo-cli init my-backend-project')}
  ${chalk_1.default.cyan('$ apollo-cli init')}
`)
    .action(async (projectName) => {
    // Prompt for Project Name if missing
    if (!projectName) {
        const answers = await inquirer_1.default.prompt([
            {
                type: 'input',
                name: 'name',
                message: 'What is the name of your project?',
                default: 'my-apollo-app',
                validate: (input) => input.trim() !== '' ? true : 'Project name is required',
            },
        ]);
        projectName = answers.name;
    }
    const projectRoot = path_1.default.join(process.cwd(), projectName);
    if (fs_extra_1.default.existsSync(projectRoot)) {
        console.error(chalk_1.default.red(`Error: Directory "${projectName}" already exists.`));
        process.exit(1);
    }
    console.log(chalk_1.default.blue(`\n🚀 Initializing project: ${chalk_1.default.bold(projectName)}...\n`));
    try {
        // Create project directory
        await fs_extra_1.default.ensureDir(projectRoot);
        // Fetch latest versions of dependencies
        const { dependencies, devDependencies } = await getLatestDependencies();
        // Create Package.json
        await (0, file_generator_1.writePackageJson)(projectRoot, projectName, dependencies, devDependencies);
        // Create tsconfig.json
        await (0, file_generator_1.writeTsConfig)(projectRoot);
        // Create scaffolding folders
        const folders = [
            'src/modules',
            'src/middlewares',
            'src/routes',
            'src/utils',
            'src/errors',
            'src/config',
            'src/lib',
        ];
        console.log(chalk_1.default.blue('📂 Creating project structure...'));
        for (const folder of folders) {
            await fs_extra_1.default.ensureDir(path_1.default.join(projectRoot, folder));
        }
        // Initialize Prisma
        console.log(chalk_1.default.blue('\n🔄 Initializing Prisma...\n'));
        try {
            (0, child_process_1.execSync)('npx prisma init --datasource-provider postgresql --output ../generated/prisma', {
                cwd: projectRoot,
                stdio: 'inherit',
            });
        }
        catch (error) {
            // Ignore if it fails (e.g., if files already exist)
        }
        // Create essential project files
        await (0, file_generator_1.writeServerFile)(projectRoot);
        await (0, file_generator_1.writeAppFile)(projectRoot);
        await (0, file_generator_1.writeConfigFile)(projectRoot);
        await (0, file_generator_1.writeEnvFile)(projectRoot);
        await (0, file_generator_1.writePrismaSchema)(projectRoot);
        await (0, file_generator_1.writeGitIgnore)(projectRoot);
        await (0, file_generator_1.writePrismaFile)(projectRoot);
        console.log(chalk_1.default.green('\n✅ Project structure created successfully! ✨'));
        // Prompt for package manager selection and installation
        const { available, all } = detectPackageManagers();
        const pmChoices = all.map((pm) => ({
            name: available.includes(pm) ? `${pm} ✓` : `${pm} (not installed)`,
            value: pm,
        }));
        const answers = await inquirer_1.default.prompt([
            {
                type: 'list',
                name: 'packageManager',
                message: chalk_1.default.blue('Which package manager would you like to use?'),
                choices: pmChoices,
                default: available[0] || 'npm',
            },
            {
                type: 'confirm',
                name: 'install',
                message: chalk_1.default.blue('Would you like to install dependencies now?'),
                default: true,
            },
        ]);
        const packageManager = answers.packageManager;
        const { install } = answers;
        // Check if selected package manager is installed
        if (install && !isPackageManagerInstalled(packageManager)) {
            console.log(chalk_1.default.yellow(`\n⚠️  ${packageManager} is not installed on your system.`));
            console.log(chalk_1.default.cyan(`Install it with: npm install -g ${packageManager}`));
            console.log(chalk_1.default.cyan(`Falling back to npm...\n`));
            const installCmd = 'npm install';
            (0, child_process_1.execSync)(installCmd, { cwd: projectRoot, stdio: 'inherit' });
            console.log(chalk_1.default.green('\n✅ Dependencies installed! ✨'));
            const devCmd = 'npm run dev';
            console.log(chalk_1.default.cyan(`\nTo get started:\n  cd ${projectName}\n  ${devCmd}\n`));
        }
        else if (install) {
            console.log(chalk_1.default.yellow(`\n📦 Installing dependencies with ${packageManager}...\n`));
            const installCmd = packageManager === 'yarn' ? 'yarn' : `${packageManager} install`;
            (0, child_process_1.execSync)(installCmd, { cwd: projectRoot, stdio: 'inherit' });
            console.log(chalk_1.default.green('\n✅ Dependencies installed! ✨'));
            const devCmd = packageManager === 'npm'
                ? 'npm run dev'
                : packageManager === 'yarn'
                    ? 'yarn dev'
                    : 'pnpm dev';
            console.log(chalk_1.default.cyan(`\nTo get started:\n  cd ${projectName}\n  ${chalk_1.default.bold(devCmd)}\n`));
        }
        else {
            const devCmd = packageManager === 'npm'
                ? 'npm run dev'
                : packageManager === 'yarn'
                    ? 'yarn dev'
                    : 'pnpm dev';
            const installCmd = packageManager === 'yarn' ? 'yarn' : `${packageManager} install`;
            console.log(chalk_1.default.cyan(`\nTo get started:\n  cd ${projectName}\n  ${chalk_1.default.bold(`${installCmd} && ${devCmd}`)}\n`));
        }
    }
    catch (error) {
        console.error(chalk_1.default.red('\n❌ Initialization failed:'), error);
        process.exit(1);
    }
});
