// @bayajidalam/apollo-cli/dist/bin/index.js
#!/usr/bin/env node
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const commander_1 = require("commander");
const build_1 = require("../commands/build");
const generate_1 = require("../commands/generate");
const init_1 = require("../commands/init");
const prisma_1 = require("../commands/prisma");
// Set up the main CLI program
const program = new commander_1.Command();
const chalk_1 = __importDefault(require("chalk"));
const figlet_1 = __importDefault(require("figlet"));
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
// Read package.json to get version and description
const packageJsonPath = path_1.default.join(__dirname, '../../package.json');
const packageJson = JSON.parse(fs_1.default.readFileSync(packageJsonPath, 'utf-8'));
program
    .version(packageJson.version)
    .description('Apollo CLI Tool')
    .addHelpText('beforeAll', chalk_1.default.cyan(figlet_1.default.textSync('Apollo CLI', { horizontalLayout: 'full' })) +
    '\n')
    .configureHelp({
    subcommandTerm: (cmd) => chalk_1.default.magenta.bold(cmd.name()),
    subcommandDescription: (cmd) => chalk_1.default.white(cmd.description()),
    optionTerm: (option) => chalk_1.default.green(option.flags),
    optionDescription: (option) => chalk_1.default.white(option.description),
    commandUsage: (cmd) => chalk_1.default.yellow(cmd.usage()),
    commandDescription: (cmd) => chalk_1.default.blue(cmd.description()),
})
    .addHelpText('after', `
${chalk_1.default.yellow('Global Examples:')}
  ${chalk_1.default.cyan('$ apollo-cli init my-project')}
  ${chalk_1.default.cyan('$ apollo-cli generate module user')}
  ${chalk_1.default.cyan('$ apollo-cli prisma migrate')}
  ${chalk_1.default.cyan('$ apollo-cli build')}
`);
// Register available commands
program.addCommand(generate_1.generateCommand);
program.addCommand(build_1.buildCommand);
program.addCommand(prisma_1.prismaCommand);
program.addCommand(init_1.initCommand);
// Parse command line arguments
program.parse(process.argv);
