// @bayajidalam/apollo-cli/dist/utils/file-generator.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.renderTemplate = renderTemplate;
exports.writePackageJson = writePackageJson;
exports.writeTsConfig = writeTsConfig;
exports.writeServerFile = writeServerFile;
exports.writeAppFile = writeAppFile;
exports.writeConfigFile = writeConfigFile;
exports.writeEnvFile = writeEnvFile;
exports.writePrismaFile = writePrismaFile;
exports.writePrismaSchema = writePrismaSchema;
exports.writeGitIgnore = writeGitIgnore;
const chalk_1 = __importDefault(require("chalk"));
const ejs_1 = __importDefault(require("ejs"));
const fs_extra_1 = __importDefault(require("fs-extra"));
const path_1 = __importDefault(require("path"));
/**
 * Renders an EJS template and writes the output to the target path.
 */
async function renderTemplate(templateName, targetPath, data = {}) {
    try {
        const templateDir = path_1.default.resolve(__dirname, '../templates');
        let templatePath = path_1.default.join(templateDir, templateName);
        // Fallback for development (src/templates)
        if (!fs_extra_1.default.existsSync(templatePath)) {
            templatePath = path_1.default.resolve(__dirname, '../../src/templates', templateName);
        }
        if (!fs_extra_1.default.existsSync(templatePath)) {
            console.error(chalk_1.default.red(`❌ Template not found at: ${templatePath}`));
            return;
        }
        const templateContent = await fs_extra_1.default.readFile(templatePath, 'utf-8');
        const renderedContent = ejs_1.default.render(templateContent, data);
        await fs_extra_1.default.ensureDir(path_1.default.dirname(targetPath));
        await fs_extra_1.default.writeFile(targetPath, renderedContent);
        console.log(chalk_1.default.green(`✓ Created: ${path_1.default.relative(process.cwd(), targetPath)}`));
    }
    catch (error) {
        console.error(chalk_1.default.red('❌ Error generating file:'), error);
        process.exit(1);
    }
}
/**
 * Generates the package.json file for the new project.
 */
async function writePackageJson(projectRoot, projectName, dependencies, devDependencies) {
    const packageJson = {
        name: projectName,
        version: '1.0.0',
        description: 'Backend project generated by Apollo Gears CLI',
        main: 'dist/server.js',
        scripts: {
            start: 'node dist/server.js',
            dev: 'ts-node-dev --respawn --transpile-only src/server.ts',
            build: 'tsc',
            lint: 'eslint src/**/*.ts',
            'lint:fix': 'eslint src/**/*.ts --fix',
            format: 'prettier --write .',
        },
        keywords: ['express', 'prisma', 'typescript', 'apollo-gears'],
        author: '',
        license: 'ISC',
        dependencies,
        devDependencies,
    };
    await fs_extra_1.default.writeJson(path_1.default.join(projectRoot, 'package.json'), packageJson, {
        spaces: 2,
    });
}
/**
 * Generates the tsconfig.json file with TypeScript configuration.
 */
async function writeTsConfig(projectRoot) {
    const tsconfig = {
        compilerOptions: {
            target: 'ES2021',
            module: 'commonjs',
            moduleResolution: 'node',
            outDir: './dist',
            rootDir: './src',
            strict: true,
            esModuleInterop: true,
            skipLibCheck: true,
            forceConsistentCasingInFileNames: true,
        },
        include: ['src/**/*'],
        exclude: ['node_modules', 'dist', 'generated/prisma'],
    };
    await fs_extra_1.default.writeJson(path_1.default.join(projectRoot, 'tsconfig.json'), tsconfig, {
        spaces: 2,
    });
}
async function writeServerFile(projectRoot) {
    await renderTemplate('init/server.ejs', path_1.default.join(projectRoot, 'src/server.ts'), { port: 5000 });
}
async function writeAppFile(projectRoot) {
    await renderTemplate('init/app.ejs', path_1.default.join(projectRoot, 'src/app.ts'));
}
async function writeConfigFile(projectRoot) {
    await renderTemplate('init/config.ejs', path_1.default.join(projectRoot, 'src/config/index.ts'));
}
async function writeEnvFile(projectRoot) {
    await renderTemplate('init/env.ejs', path_1.default.join(projectRoot, '.env'));
}
async function writePrismaFile(projectRoot) {
    await renderTemplate('init/prisma.ejs', path_1.default.join(projectRoot, 'src/lib/prisma.ts'));
}
async function writePrismaSchema(projectRoot) {
    await renderTemplate('init/prisma-schema.ejs', path_1.default.join(projectRoot, 'prisma/schema.prisma'));
}
async function writeGitIgnore(projectRoot) {
    await renderTemplate('init/gitignore.ejs', path_1.default.join(projectRoot, '.gitignore'));
}
